### DISCUSSION TITLE
DataGrid - scroll bar cuts content

### DISCUSSION THREAD

**Discussion by KrzysztofDusko** [AvaloniaUI/Avalonia#14445] [Q&A]
Simmilar to https://github.com/AvaloniaUI/Avalonia/issues/12182
and to https://github.com/AvaloniaUI/Avalonia/discussions/11726
If we set AllowAutoHide to false in datagrid ScrollBar selector in this way 
```
				<Style Selector="ScrollBar">
					<Setter Property="Background" Value="Blue"/>
					<Setter Property="AllowAutoHide" Value="False"/>
				</Style>
```
last row is not visible..  for demonstration i used  very big ScrollBarSize
```
		<DataGrid.Resources>
			<x:Double x:Key="ScrollBarSize">20.0</x:Double>
		</DataGrid.Resources>
```
![image](https://github.com/AvaloniaUI/Avalonia/assets/69449360/e7e88a52-0b4b-47a5-988c-aa647071b08f)

i found than I can achieve this effect by editing Fluent.xaml in Avalonia.Controls.DataGrid/Themes
![image](https://github.com/AvaloniaUI/Avalonia/assets/69449360/67cd0f6e-4f94-46e9-b453-b627b6af468f)
by changing Grid.RowSpan/Grid.ColumnSpan
![image](https://github.com/AvaloniaUI/Avalonia/assets/69449360/50c962de-6d47-4a18-9ab3-56c022c741ac)
from 2,3 to respectivly 1,2 is there any way to ovveride this values ? 
https://github.com/AvaloniaUI/Avalonia/blob/e0fb7db6c56f745027ef1b041ea0ca24feb27ab5/src/Avalonia.Controls.DataGrid/Themes/Fluent.xaml#L538

i tried this.. Color works but row/column spans not :(
```
				<Style Selector="DataGridRowsPresenter#PART_RowsPresenter">
					<Setter Property="Grid.RowSpan" Value="1"/>
					<Setter Property="Grid.ColumnSpan" Value="2"/>
					<Setter Property="Background" Value="Green"/>
				</Style>	
```

**Comment by thevortexcloud** [comment]
You can override the control template to make it look like what ever you want it to look like.

**Comment by timunie** [comment]
Should be available for DataGrid as well. https://github.com/AvaloniaUI/Avalonia/pull/12787 

If you use a custom theme, make sure to have the same fix applied.

**Reply by KrzysztofDusko** [reply]
thanks for answer but in my opinion AllowAutoHide is set to false
```xml
<Style Selector="DataGrid /template/ ScrollBar#PART_HorizontalScrollbar">
    <Setter Property="AllowAutoHide" Value="False"/>
</Style>
```
this should also change default scroll bar position.. Right now when scroll bar is viible with no auto hide last row is no full visible. Right now default behavior is wrong..
(standard fulent theme)
![image](https://github.com/AvaloniaUI/Avalonia/assets/69449360/b46d5d25-574e-4597-b8dd-13f6cc0aa931)

**Reply by KrzysztofDusko** [reply]
like there  Grid.RowSpan / Grid.ColumnSpan should not be hardcoded..
https://github.com/AvaloniaUI/Avalonia/pull/12301/commits/24c7398f240b0709577f412108ef76e99422ef57

**Reply by timunie** [reply]
It shouldn't be hardcoded in latest master. 🤔

**Reply by KrzysztofDusko** [reply]
well it is ? in ScrollViewer is OK in data grid not. And fix in scrollviewer only cant be applied there.. 
https://github.com/AvaloniaUI/Avalonia/blob/bee1c045812413969e5e13e1648b880648dfcffd/src/Avalonia.Controls.DataGrid/Themes/Fluent.xaml#L535-L540

i tried to do it myself but i cant write valid selecotr for this :(
```xaml
				<Style Selector="^(ScrollBar#PART_HorizontalScrollbar.AllowAutoHide)=False /template/ DataGridRowsPresenter#PART_RowsPresenter">
					 .... 
				</Style>
```
of course my issues is fixed but i copied whole DG style and edit ... 
In this way:
1. remove hardcoded values 
2. set in DataGrid.Styles (i want this only for horizontal scroll bar)
```xaml
<Style Selector="DataGrid /template/ ScrollBar#PART_HorizontalScrollbar">
    <Setter Property="AllowAutoHide" Value="False"/>
</Style>
<Style Selector="DataGrid /template/ ScrollBar#PART_VerticalScrollbar">
    <Setter Property="AllowAutoHide" Value="True"/>
</Style>					
<Style Selector="DataGrid /template/ DataGridRowsPresenter#PART_RowsPresenter">
    <Setter Property="Grid.RowSpan" Value="1" />
   <Setter Property="Grid.ColumnSpan" Value="3" />
</Style>	
```

**Reply by timunie** [reply]
Looks like a fix I thought I made wasn't made in reality. Not on upstream Avalonia at least... So we need to update that in source first. Hardcoded setters must be removed by a Style selector. Will make an issue for it. If you want to have it quickly done, a PR is welcome.

Sorry for the confusion 💐
